File: python/exer03a_readers_writers.py
'''
THE READERS-WRITERS PROBLEM
Solution for the first readers-writers problem
'''

import random
import time
import threading



class GlobalData:
    def __init__(self):
        self.resource = 0
        self.reader_count = 0
        self.lk_resource = threading.Lock()
        self.lk_reader_count = threading.Lock()



def do_task_writer(g: GlobalData, time_delay: int):
    time.sleep(time_delay)

    with g.lk_resource:
        g.resource = random.randint(0, 99)
        print('Write', g.resource)



def do_task_reader(g: GlobalData, time_delay: int):
    time.sleep(time_delay)

    # Increase reader count
    with g.lk_reader_count:
        g.reader_count += 1

        if g.reader_count == 1:
            g.lk_resource.acquire()

    # Do the reading
    print('Read', g.resource)

    # Decrease reader count
    with g.lk_reader_count:
        g.reader_count -= 1

        if g.reader_count == 0:
            g.lk_resource.release()



gbl_data = GlobalData()
NUM_READERS = 8
NUM_WRITERS = 6

for _ in range(NUM_READERS):
    threading.Thread(target=do_task_reader, args=(gbl_data, random.randint(0, 2))).start()

for _ in range(NUM_WRITERS):
    threading.Thread(target=do_task_writer, args=(gbl_data, random.randint(0, 2))).start()

File: cpp/cpp-std/exer03a-readers-writers.cpp
/*
THE READERS-WRITERS PROBLEM
Solution for the first readers-writers problem
*/


#include <iostream>
#include <chrono>
#include <thread>
#include <mutex>
#include "mylib-random.hpp"
using namespace std;



struct GlobalData {
    volatile int resource = 0;
    int readerCount = 0;

    std::mutex mutResource;
    std::mutex mutReaderCount;
};



void doTaskWriter(GlobalData* g, int timeDelay) {
    std::this_thread::sleep_for(std::chrono::seconds(timeDelay));

    g->mutResource.lock();

    g->resource = mylib::RandInt::get(100);
    cout << "Write " << g->resource << endl;

    g->mutResource.unlock();
}



void doTaskReader(GlobalData* g, int timeDelay) {
    std::this_thread::sleep_for(std::chrono::seconds(timeDelay));


    // Increase reader count
    g->mutReaderCount.lock();
    g->readerCount += 1;

    if (1 == g->readerCount)
        g->mutResource.lock();

    g->mutReaderCount.unlock();


    // Do the reading
    cout << "Read " << g->resource << endl;


    // Decrease reader count
    g->mutReaderCount.lock();
    g->readerCount -= 1;

    if (0 == g->readerCount)
        g->mutResource.unlock();

    g->mutReaderCount.unlock();
}



int main() {
    GlobalData globalData;


    constexpr int NUM_READERS = 8;
    constexpr int NUM_WRITERS = 6;

    std::thread lstThReader[NUM_READERS];
    std::thread lstThWriter[NUM_WRITERS];


    // CREATE THREADS
    for (auto&& th: lstThReader) {
        th = std::thread(&doTaskReader, &globalData, mylib::RandInt::get(3));
    }

    for (auto&& th: lstThWriter) {
        th = std::thread(&doTaskWriter, &globalData, mylib::RandInt::get(3));
    }


    // JOIN THREADS
    for (auto&& th: lstThReader) {
        th.join();
    }

    for (auto&& th: lstThWriter) {
        th.join();
    }


    return 0;
}
